django/db/models/sql/query.py

```diff
@@ -2108,3 +2108,435 @@ def update_join_types(self, query):
         query.promote_joins(to_promote)
         query.demote_joins(to_demote)
         return to_demote
+
+
+class ExpressionConstraintQuery:
```

Comment:
It feels odd why so much code needs to be duplicated. Have you considered letting `Query` subclass this?